flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1286,10 +1289,38 @@ private void transferAllDataFromStateHandles(
 			Map<StateHandleID, StreamStateHandle> stateHandleMap,
 			Path restoreInstancePath) throws IOException {
 
+			ExecutorService executorService = Executors.newFixedThreadPool(5);
+			List<FutureTask<Void>> tasks = new ArrayList<>(stateHandleMap.size());
+
 			for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
 				StateHandleID stateHandleID = entry.getKey();
 				StreamStateHandle remoteFileHandle = entry.getValue();
-				copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle);
+
+				FutureTask<Void> task = new FutureTask<>(() -> {
+					try {
+						copyStateDataHandleData(new Path(restoreInstancePath, stateHandleID.toString()), remoteFileHandle);
+					} catch (IOException e) {
+						LOG.error("Copy State Data error, stateHandleID [{}], remoteFileHandle[{}]", stateHandleID.toString(), remoteFileHandle);
+						throw e;
+					}
+					return null;
+				});
+				tasks.add(task);
+				executorService.submit(task);
```

Comment:
I suggest to use `CompletableFuture.runAsync(Runnable, Executor)` instead of `FutureTask`. It allows to use `FutureUtils.waitForAll(CompletableFutures).get()`.